crates/tx-graph/src/pog_musig_functor.rs

```diff
@@ -0,0 +1,328 @@
+use std::future::Future;
+
+use futures::future::join_all;
+use serde::{Deserialize, Serialize};
+use strata_bridge_primitives::constants::NUM_ASSERT_DATA_TX;
+
+use crate::transactions::assert_chain::{deserialize_assert_vector, serialize_assert_vector};
+
+/// Functor like data structure for holding an arbitrary data structure that is matched with each of
+/// the inputs of the peg-out graph.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PogMusigF<T> {
+    /// Data associated with the challenge transaction input.
+    pub challenge: T,
+
+    /// Data associated with the pre-assert transaction input.
+    pub pre_assert: T,
+
+    /// Data associated with the post-assert transaction inputs.
+    #[serde(serialize_with = "serialize_assert_vector")]
+    #[serde(deserialize_with = "deserialize_assert_vector")]
+    pub post_assert: [T; NUM_ASSERT_DATA_TX],
+
+    /// Data associated with the payout optimistic transaction inputs.
+    pub payout_optimistic: [T; 5],
+
+    /// Data associated with the payout transaction inputs.
+    pub payout: [T; 4],
+
+    /// Data associated with the disprove transaction input.
+    pub disprove: T,
+
+    /// Data for each of the slash stake transaction input pairs.
+    pub slash_stake: Vec<[T; 2]>,
```

Comment:
Can we use constants for these 5, 4 and 2 lengths like we did in the `post_assert`?